File: huggingplace-sdk/src/session.js
import { HuggingPlace } from './huggingplace.js';

/**
 * Session class for managing related interactions
 */
export class Session {
  /**
   * Create a new session
   * @param {HuggingPlace} huggingplace - HuggingPlace instance
   * @param {string} sessionId - Session ID
   * @param {Object} options - Session options
   * @param {Object} [options.metaData] - Default metadata for the session
   * @param {Object} [options.user_meta_data] - Default user metadata for the session
   */
  constructor(huggingplace, sessionId, options = {}) {
    this.huggingplace = huggingplace;
    this.sessionId = sessionId;
    this.defaultMetadata = options.metaData || {};
    this.defaultUserMetaData = options.user_meta_data || {};
  }

  /**
   * Log an interaction with session defaults
   * @param {Object} options - Log options
   * @param {string} options.user_prompt - User prompt
   * @param {string} options.ai_response - AI response
   * @param {string} [options.session_id] - Session ID (uses session default if not provided)
   * @param {Object} [options.metaData] - Additional metadata
   * @param {Object} [options.user_meta_data] - Additional user metadata
   * @returns {Promise<void>}
   */
  async log(options) {
    // Merge session defaults with provided options
    const mergedOptions = {
      ...options,
      // Only set session_id if not provided by user
      session_id: options.session_id || this.sessionId,
      metaData: {
        ...this.defaultMetadata,
        ...options.metaData,
      },
      user_meta_data: {
        ...this.defaultUserMetaData,
        ...options.user_meta_data,
      },
    };

    await this.huggingplace.log(mergedOptions);
  }

  /**
   * Log an individual processing step
   * @param {Object} options - Step options
   * @param {string} options.type - Step type
   * @param {string} options.userQuestion - User question for this step
   * @param {string} options.promptResponse - Response for this step
   * @param {string} [options.llmModel] - LLM model used
   * @param {number} [options.token] - Token count
   * @param {number} [options.responseTime] - Response time
   * @param {number} [options.inputTokens] - Input token count
   * @param {number} [options.outputTokens] - Output token count
   * @returns {Promise<void>}
   */
  async logStep(options) {
    await this.huggingplace.logStep(options);
  }

  /**
   * Update session defaults
   * @param {Object} [metadata] - New default metadata
   * @param {Object} [userMetaData] - New default user metadata
   */
  updateDefaults(metadata, userMetaData) {
    if (metadata) {
      this.defaultMetadata = { ...this.defaultMetadata, ...metadata };
    }
    if (userMetaData) {
      this.defaultUserMetaData = { ...this.defaultUserMetaData, ...userMetaData };
    }
  }

  /**
   * Get current session information
   * @returns {Object} Session information
   */
  getSessionInfo() {
    return {
      sessionId: this.sessionId,
      defaultMetadata: this.defaultMetadata,
      defaultUserMetaData: this.defaultUserMetaData,
    };
  }
} 

File: huggingplace-python-sdk/huggingplace_sdk/session.py
"""
Session class for managing related interactions
"""

from typing import Dict, Any, Optional
from .huggingplace import HuggingPlace


class Session:
    """
    Session class for managing related interactions.
    """
    
    def __init__(
        self, 
        huggingplace: HuggingPlace, 
        session_id: str, 
        options: Optional[Dict[str, Any]] = None
    ):
        """
        Create a new session.
        
        Args:
            huggingplace: HuggingPlace instance
            session_id: Session ID
            options: Session options
        """
        self.huggingplace = huggingplace
        self.session_id = session_id
        self.default_metadata = options.get("metadata", {}) if options else {}
        self.default_user_metadata = options.get("user_metadata", {}) if options else {}

    async def log(self, options: Dict[str, Any]) -> None:
        """
        Log an interaction with session defaults.
        
        Args:
            options: Log options
                - user_prompt: User prompt
                - response: AI response
                - session_id: Session ID (uses session default if not provided)
                - metadata: Additional metadata
                - user_metadata: Additional user metadata
        """
        # Merge session defaults with provided options
        merged_options = {
            **options,
            # Only set session_id if not provided by user
            "session_id": options.get("session_id") or self.session_id,
            "metadata": {
                **self.default_metadata,
                **options.get("metadata", {}),
            },
            "user_metadata": {
                **self.default_user_metadata,
                **options.get("user_metadata", {}),
            },
        }

        await self.huggingplace.log(merged_options)

    async def log_step(self, options: Dict[str, Any]) -> None:
        """
        Log an individual processing step.
        
        Args:
            options: Step options
                - type: Step type
                - user_question: User question for this step
                - prompt_response: Response for this step
                - llm_model: LLM model used
                - token: Token count
                - response_time: Response time
                - input_tokens: Input token count
                - output_tokens: Output token count
        """
        await self.huggingplace.log_step(options)

    def update_defaults(
        self, 
        metadata: Optional[Dict[str, Any]] = None, 
        user_metadata: Optional[Dict[str, Any]] = None
    ) -> None:
        """
        Update session defaults.
        
        Args:
            metadata: New default metadata
            user_metadata: New default user metadata
        """
        if metadata:
            self.default_metadata = {**self.default_metadata, **metadata}
        if user_metadata:
            self.default_user_metadata = {**self.default_user_metadata, **user_metadata}

    def get_session_info(self) -> Dict[str, Any]:
        """
        Get current session information.
        
        Returns:
            Session information
        """
        return {
            "session_id": self.session_id,
            "default_metadata": self.default_metadata,
            "default_user_metadata": self.default_user_metadata,
        } 